### ISSUE TITLE
Race condition in task_remover leads to increasing stale task logs and disk usage

### ISSUE THREAD

**Issue by aegershman** [cloudfoundry/bosh#2232]
**Describe the bug**

TL;DR; when API calls are made to the `bosh vms` or `bosh is` endpoints in parallel, the file count of `/var/vcap/store/director/tasks/` will tend to increase over time, far above the expected stable number of ~2000. These files (task logs) consumes persistent disk and slows down file operations.

**Describe the bug: extended edition**

Please see this slack thread for the summary of context and identified problemspace: https://cloudfoundry.slack.com/archives/C02HPPYQ2/p1579885333011900

The [`prometheus-boshrelease`](https://github.com/bosh-prometheus/prometheus-boshrelease) is used extensively. It pulls metrics from deployments the director is responsible for through the [`monitor-bosh`](https://github.com/bosh-prometheus/prometheus-boshrelease/blob/master/manifests/operators/monitor-bosh.yml) operations file, which leverages the [`bosh_exporter`](https://github.com/bosh-prometheus/bosh_exporter) to scrape. In turn, each scrape [triggers the `bosh_exporter` to request instance info](https://github.com/bosh-prometheus/bosh_exporter/blob/master/collectors/bosh_collector.go) for each deployment under the director, which kicks off a number of `bosh instances` requests to the director, all at once. If you’re leveraging the prometheus boshrelease like this, you can see this in action when you do `bosh tasks —recent —all`, and see something similar to this, where all these “retrieve vm-stats” requests are coming in in groupings which start at basically the same time, with each request being handled by different threads.  

```
 20654561  done   Sun Jan 26 20:21:54 UTC 2020  Sun Jan 26 20:21:54 UTC 2020  bosh_exporter  service-instance_15001beb-4db9-4227-bea5-e5b0fc0dd98e  retrieve vm-stats  -
20654560  done   Sun Jan 26 20:21:54 UTC 2020  Sun Jan 26 20:21:54 UTC 2020  bosh_exporter  p-rabbitmq-d31675f4e051a9411c20                        retrieve vm-stats  -
20654559  done   Sun Jan 26 20:21:54 UTC 2020  Sun Jan 26 20:21:54 UTC 2020  bosh_exporter  service-instance_31aa22dd-50af-4a44-a62a-a32364cafe7c  retrieve vm-stats  -
20654558  done   Sun Jan 26 20:21:54 UTC 2020  Sun Jan 26 20:21:54 UTC 2020  bosh_exporter  yugabyte_service_broker-6bf9a63a53399cbdd2dc           retrieve vm-stats  -
20654557  done   Sun Jan 26 20:21:54 UTC 2020  Sun Jan 26 20:21:54 UTC 2020  bosh_exporter  service-instance_ebf11fb7-241e-44ef-b9b3-4cb5b89ba0f9  retrieve vm-stats  -
20654556  done   Sun Jan 26 20:21:52 UTC 2020  Sun Jan 26 20:21:53 UTC 2020  bosh_exporter  concourse-apps                                         retrieve vm-stats  -
20654555  done   Sun Jan 26 20:21:49 UTC 2020  Sun Jan 26 20:21:49 UTC 2020  bosh_exporter  p_spring-cloud-services-b4f086e24529af4a55d5           retrieve vm-stats  -
20654554  done   Sun Jan 26 20:21:52 UTC 2020  Sun Jan 26 20:21:52 UTC 2020  bosh_exporter  service-instance_bdba0627-ee4f-46cd-a8ea-6ea2453866df  retrieve vm-stats  -
20654553  done   Sun Jan 26 20:21:54 UTC 2020  Sun Jan 26 20:21:54 UTC 2020  bosh_exporter  p-event-alerts-755f22806f19d1735230                    retrieve vm-stats  -
20654552  done   Sun Jan 26 20:21:53 UTC 2020  Sun Jan 26 20:21:53 UTC 2020  bosh_exporter  prometheus                                             retrieve vm-stats  -
20654551  done   Sun Jan 26 20:21:54 UTC 2020  Sun Jan 26 20:21:54 UTC 2020  bosh_exporter  service-instance_128517e7-6759-416a-8cbb-b388e00507ae  retrieve vm-stats  -
20654550  done   Sun Jan 26 20:21:52 UTC 2020  Sun Jan 26 20:21:53 UTC 2020  bosh_exporter  p-spring-cloud-services-ed246990d486488edc57           retrieve vm-stats  -
20654549  done   Sun Jan 26 20:21:49 UTC 2020  Sun Jan 26 20:21:49 UTC 2020  bosh_exporter  pivotal-mysql-6ce1a7ddcbd465a04fc4                     retrieve vm-stats 
```

This is where things get exciting, and where it’s theorizing, but hear it out— basically, it’s a race condition on deleting the same set of old tasks.

Because all these requests are coming in at practically the exact same time, they’re being handled by different threads. As part of the `bosh {vms,instances}` api flow, there will be 10 old task logs selected for removal. The idea being, as new requests come in, the old tasks will be cleaned up as things go. But if we take a peek at the [`task_remover.rb` code, we see that the same set of 10 tasks can be selected as candidates for removal by all threads at the same time](https://github.com/cloudfoundry/bosh/blob/983b1ba4e14217ac3a0554b56cd144f46810fa51/src/bosh-director/lib/bosh/director/api/task_remover.rb#L19), and only one thread will ‘commit’ the actual removal. 

>it's possible for multiple threads to initiate task removal
>both could get the same results from removal_candidates_dataset,
>but only the first would succeed at deletion; ignore failure of
>subsequent attempts

After those 10 removal candidates are removed by the first thread to complete, the other threads _which had selected those same threads_ (or at least some of them, I guess) will try to delete them, but will ignore the fact those threads were already deleted and move on with their lives. This causes the total count of `vms`-tasks to increase over time (our env of ~2 years was at 4.7 million task `vms` logs. whew.)

**To Reproduce**

Deploy a BOSH director
Deploy bosh-prometheus/prometheus-boshrelease leveraging the `monitor-bosh` operator, which in-turn leverages the bosh_exporter
Observe the number `vms`-task logs increase over 2000 over time as part of each bosh_exporter scrape

You can check your director's task count:

```sh
/var/vcap/jobs/director/bin/console
Bosh::Director::Models::Task.group_and_count(:type).all.map {|a| "type: #{a[:type]} count: #{a[:count]}" }
```

**Expected behavior**
The count of `vms`-tasks to stay at the expected stable rate of ~2,000

**Versions (please complete the following information):**
 - Infrastructure: AWS
 - BOSH version: 270.4.2
 - BOSH CLI version: 6.2.0-e5ce64ab-2020-01-16T19:22:06Z
 - Stemcell version: 456.74

See also:
- https://cloudfoundry.slack.com/archives/C02HPPYQ2/p1579885333011900
- https://github.com/bosh-prometheus/bosh_exporter/issues/22#issuecomment-578539779
- https://github.com/cloudfoundry/bosh/issues/1871#issuecomment-480278657
- https://www.pivotaltracker.com/n/projects/956238/stories/163150850

We have a mitigation in place (+1 hax4noodles, mfine, etc., thank you), however the fundamental causes of the problem will persist.

Thanks all for your help and the work you all do. Let me know if I can be of more assistance.

**Comment by voelzmo** [comment]
Thanks for the awesome bug report and thorough write-up!

**Comment by xtreme-conor-nosal** [comment]
Another concern is that TaskRemover is invoked when the task is queued, and is only concerned with terminated tasks (not processing or queued). So if there are 2000 completed tasks, and you queue up 100, TaskRemover won't do anything. Then it will take 12 serial tasks to shrink the history back to 2000.

The number of queued/processing tasks should be taken into consideration and dynamically decide how many to delete instead of hardcoding 10, and cleanup can be extracted from the JobQueue class into its own scheduled task.

**Comment by h4xnoodle** [comment]
Current proposal for a solution:
1. Create another scheduler that can be configured by the operator similar to the other [existing schedulers](https://github.com/cloudfoundry/bosh/blob/master/jobs/director/spec#L126). Scheduling allows operators to clean tasks at their leisure. Default will be once a week.
1. Enqueue a 'clean up tasks' job, that will use the existing `TaskRemover` to clean tasks. One worker will take this job, removing the race condition.
1. Modify the task remover to remove as many tasks necessary to remain at `max_tasks` instead of hardcoded value of 10. This will 'fix' the existing directors in the state of millions of tasks. The task log files will also be deleted, as per normal.

Let me know if there are potential problems with this!
